Correcting ordinary insert statement for tables with composite primary keys

trueqbit opened the pull request:
### Improvements:
* Objects mapped to tables with composite primary keys are now insertable with an ordinary insert statement: they get inserted with the values of the specified object except for default values.
  * This improvement contains a bug fix for insert statements into tables without rowid where providing values for primary key columns is required (and expected).
  
### Other improvements:
* Validate base table definition when calling `make_table()`.
  * Specifically this ensures that only one primary key definition is specified.
  * It also means that we can get rid of a runtime error when serializing a `get` query statement.
* Validate column definition with respect to auto-increment specifier on a column primary key and the field type.
* Improved documentation as a comment on the ordinary 'insert' statement into tables, describing how the routine behaves with rowid vs. without rowid tables and with a single-column primary key vs. composite keys.
* Many more unit tests of course.

### Fixes
* Corrected some SQLite version test that enable version-dependent SQLite features.

trueqbit committed:
Fixed handling composite primary key columns for tables without rowid

The if-condition used for filtering primary key columns must be the same at run-time as at compile-time.

trueqbit committed:
Unit tests for insertions into tables without rowid and compound PKs

trueqbit committed:
Check up front whether type of range to insert or replace is convertible

... to the mapped object type, avoiding hard to read error messages later.

trueqbit committed:
Removed the period after static assert messages, according to the prevailing style

trueqbit committed:
Unit test for serializing for insert statements involving a primary key

trueqbit committed:
Moved cpp guard that enables testing projected range insertions

trueqbit committed:
Same compiler warning options for GNU as for Clang for unit tests

trueqbit committed:
Renamed functions dealing with the table primary key

Up until this commit, the functions were referred to as "composite PK functions", although this actually always meant "table PK functions", regardless of whether they were single or composite primary keys defined at the table level.

trueqbit committed:
Omit only single primary key columns in insert statements

trueqbit committed:
Renamed auxiliary function testing whether a primary key contains a column

trueqbit committed:
Addressed a compiler error that occurred with msvc 141

trueqbit committed:
Validate primary key requirements up front

trueqbit committed:
Improved filter for excluding columns during serialization

* Improved understanding through legibility.

trueqbit committed:
Returned the properly typed last inserted rowid

For proper type safety `storage_t<>::insert()` needs to return an `sqlite3_int64` instead of an `int`.
This is the best way to correct the problem. Usually, compilers just accept narrowing type conversions, the nicest thing that can happen is a warning message, the worst is an error depending on compiler flags. The programmer has to handle the rest and it is the best thing to do anyway.

trueqbit committed:
Improved column primary key checks

* Check column definition up front when making the column.
* Renamed internal traits.
* Removed unit test file `is_column_with_insertable_primary_key.cpp`. It became superfluous during the course of development in the past years. Column constraints are checked elsewhere already, and it is now only a duplicate of `is_primary_key_insertable.cpp`.

trueqbit committed:
Deprecated types other than 64-bit signed integer for autoincrement PKs

SQLite documentation clearly states that an auto-incrementable primary key column must be a 64-bit signed integer type.

trueqbit committed:
Ordinary insert statement skips primary key columns with a default value

This approach keeps the previous way if all composite primary key columns have a default value, yet still enables inserting records if the table has composite primary key columns with only some or no default values.

trueqbit committed:
Fixed pre-C++23 if constexpr expression

trueqbit committed:
Corrected CPP guards for some SQLite features

* WITHOUT ROWID is only available since SQLite 3.8.2.
* FTS5: consider that it can be explicitly enabled with `SQLITE_ORM_ENABLE_FTS5`.
* Auxiliary virtual table columns are only available since SQLite 3.24.0.
* Hidden virtual table columns have no version attached.

trueqbit committed:
Taking into account that the rowid key might represented by a custom type

trueqbit committed:
Deprecated types other than 64-bit signed integer for single-column PKs

Only types that can represent a 64-bit signed integer - either integral or custom types - have the required properties to be an alias for the rowid and can cover the whole value range of it.

fnc12 commented on dev/schema/column.h:
```diff
@@ -84,6 +85,14 @@ namespace sqlite_orm {
             }
 
+            /**
+             *  Checks whether contraints contain specified class template.
```
typo: con[s]traints

fnc12 commented on dev/schema/column.h:
```diff
@@ -165,19 +174,65 @@ namespace sqlite_orm {
             filter_tuple_sequence_t<Elements, mpl::disjunction_fn<is_column, is_hidden_column>::template fn>>;
 
-#if SQLITE_VERSION_NUMBER >= 3031000
+        // Custom type: programmer's responsibility to garantee data integrity in the value range of a 64-bit signed integer
```
typo: g[u]arantee

fnc12 commented on dev/schema/table.h:
```diff
@@ -117,6 +115,25 @@ namespace sqlite_orm {
+                constexpr size_t nTablePrimaryKeyColumns =
+                    nested_tuple_size_for_t<columns_tuple_t, elements_type, pk_index_sequence>::value;
+
+                static_assert(nTablePrimaryKeyColumns > 0, "Tabel primary key definition must contain one column");
```
typo: `Table` not `Tabel`

fnc12 commented on dev/storage.h:
```diff
@@ -928,24 +939,33 @@ namespace sqlite_orm {
 
             template<class O, class... Cols>
-            int insert(const O& o, columns_t<Cols...> cols) {
+            int64 insert(const O& o, columns_t<Cols...> cols) {
```
on man, it will produce a lot of warnings. I'd leave it like this just like it is done in Java (where almost all integer operations are handles by simple `int`s) for simplicity and if somebody wants to get more precise `int64` result one can use `last_insert_rowid()` instead or use a prepared statement of `insert`. I mean this function `int insert` is a real legacy and I'd leave its signature like this just like the below one's

trueqbit replied:
OK!

fnc12 commented on examples/except_intersection.cpp:
```diff
@@ -9,12 +9,12 @@ using std::cout;
 
 struct DeptMaster {
-    int deptId = 0;
+    sqlite_orm::int64 deptId = 0;
```
what happens to this example if we skip this change?

trueqbit replied:
Does it hurt if we keep educating programmers to use `int64`? On my platforms int is 32-bit.
But I can revert, and would need to do so for all examples.

fnc12 replied:
no I mean I really want you to answer the question so I can ensure I understood your changes right. It is totally ok to educate programmers to do right things.

trueqbit replied:
Oh ok. The reason I changed it is to use the datatype that SQLite uses and assumes for the [rowid](https://sqlite.org/autoinc.html).

fnc12 replied:
what happens to this example if we skip this change? will it compile? will the warning appear?

trueqbit replied:
The example will compile with both types - `int` and `int64` - without warnings. As discussed there will be no deprecation warning for integral types other than `int64`, however `int64` is still the most accurate type for the rowid alias.

fnc12 replied:
got it, thank you. Now I understand it quite correct =)

fnc12 commented on tests/prepared_statement_tests/replace_range.cpp:
```diff
@@ -75,11 +75,19 @@ TEST_CASE("Prepared replace range") {
             REQUIRE(get<1>(statement) == userPointers.end());
             storage.execute(statement);
         }
+        SECTION("references") {
```
this type of sections is a really good shot

fnc12 commented on tests/tests5.cpp:
```diff
@@ -315,9 +315,9 @@ TEST_CASE("issue822") {
                                            primary_key(&A::getAddress, &A::getType, &A::getIndex)));
     storage.sync_schema();
-    storage.replace(A(1, 1, 0, std::make_shared<double>(55.5)));
+    storage.insert(A(1, 1, 0, std::make_shared<double>(55.5)));
```
this is a really good bug fix

trueqbit committed:
Addressed compilation error with msvc 141

trueqbit committed:
More [[maybe_unused]] arguments

trueqbit committed:
Corrected typos in comments

trueqbit committed:
Allows non-64-bit integral types as rowid alias again

Removes the deprecation notice for using non-64-bit integral types as rowid aliases, deferring the type size integrity to the programmer. There are valid concerns that deprecating integral types with less precision than `sqlite3_int64` will provoke situations where people will start complaining about many compiler warnings.

Note: I have retained the partial template specializations for validating integral types for better understanding.

trueqbit committed:
Reverts `insert` return type to `int`

There are valid concerns that deprecating integral types with less precision than `sqlite3_int64` will lead to people complaining about many warnings. As the ordinary insert routine is considered a bit like 'legacy' and is subject to multi-threading issues as well the decision is to implement another object insertion routine in the future and just leave things as they are.

fnc12 commented on dev/storage.h:
```diff
@@ -244,17 +254,16 @@ namespace sqlite_orm {
+                        "While SQLite allows primary keys of any type, sqlite_orm restricts an ordinary 'insert' into "
+                        "tables with single-column primary keys to those with an implicitly insertable column because "
+                        "it is the 'rowid' alias or has a default value."
+                        "Instead, please use 'replace' or 'insert' with explicitly specified columns.");
```
nit: I'd replace 
```
'insert' with explicitly specified columns.
```
with
```
insert(object, columns(...));
```
just to make it more recognizable cause people usually understand examples more likely than definitions

fnc12 commented on dev/storage.h:
```diff
@@ -927,6 +939,14 @@ namespace sqlite_orm {
+             *  @return The ID of the last inserted record for a rowid table, otherwise a meaningless value.
+             *          Attention: `sqlite3_last_insert_rowid()` is used to retrieve the last inserted ID, therefore the ID is only useful in single-threaded contexts.
+             *          Attention: While SQLite returns a 64-bit integer as rowid, this function returns an `int` that most likely has less precision.
+             *                     If you need the full 64-bit rowid, use `storage_t<>::execute()` instead.
```
"...or use `last_insert_rowid()` function call manually"
(and the same to the comment below)

fnc12 commented on dev/storage.h:
```diff
@@ -1530,17 +1564,25 @@ namespace sqlite_orm {
+                    using without_rowid = typename table_type::is_without_rowid;
+
+                    table.template for_each_column_excluding<  ///
+                        mpl::disjunction<
```
nit: probably this could be more readable with local aliases like
```c++
using mpl_or = mpl::disjunction;
using mpl_and = mpl::conjunction;
using mpl_not = mpl::not_;
```

trueqbit replied:
Well, if this is a serious wish, then we should flatten the namespaces and just define everything in terms of `mpl_` prefixes...

fnc12 replied:
it is not so serious wish, it is just a nit, cause you're more familiar with mpl part of the lib

trueqbit committed:
Made documentation just a bit clearer

trueqbit committed:
Simplified expression to exclude columns at compile-time
